src/lighteval/models/openai_model.py

```diff
@@ -0,0 +1,241 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import os
+import time
+from concurrent.futures import ThreadPoolExecutor
+from typing import Optional
+
+from tqdm import tqdm
+
+from lighteval.data import GenerativeTaskDataset, LoglikelihoodDataset
+from lighteval.logging.hierarchical_logger import hlog_warn
+from lighteval.models.abstract_model import LightevalModel
+from lighteval.models.endpoint_model import ModelInfo
+from lighteval.models.model_output import (
+    GenerativeResponse,
+    LoglikelihoodResponse,
+    LoglikelihoodSingleTokenResponse,
+)
+from lighteval.tasks.requests import (
+    GreedyUntilRequest,
+    LoglikelihoodRequest,
+    LoglikelihoodRollingRequest,
+    LoglikelihoodSingleTokenRequest,
+)
+from lighteval.utils.imports import is_openai_available
+
+
+if is_openai_available():
+    import logging
+
+    import tiktoken
+    from openai import OpenAI
+
+    logging.getLogger("openai").setLevel(logging.ERROR)
+    logging.getLogger("httpx").setLevel(logging.ERROR)
+
+
+class OpenAIClient(LightevalModel):
+    _DEFAULT_MAX_LENGTH: int = 4096
+
+    def __init__(self, config, env_config) -> None:
+        api_key = os.environ["OPENAI_API_KEY"]
+        self.client = OpenAI(api_key=api_key)
+
+        self.model_info = ModelInfo(
+            model_name=config.model,
+            model_sha="",
+            model_dtype=None,
+            model_size="",
+        )
+        self.API_MAX_RETRY = 5
+        self.API_RETRY_SLEEP = 3
+        self.CONCURENT_CALLS = 100
+        self.model = config.model
+        self._tokenizer = tiktoken.encoding_for_model(self.model)
+        self.pairwise_tokenization = False
+
+    def __call_api(self, prompt, return_logits, max_new_tokens, num_samples, logit_bias):
+        for _ in range(self.API_MAX_RETRY):
+            try:
+                response = self.client.chat.completions.create(
+                    model=self.model,
+                    messages=[{"role": "user", "content": prompt}],
+                    response_format={"type": "text"},
+                    max_tokens=max_new_tokens if max_new_tokens > 0 else 100,
+                    logprobs=return_logits,
+                    top_logprobs=3,
```

Comment:
Does top_logprob induce some form of sampling? - we don't want it for all cases

Reply:
It does not, but we do not need it

Reply:
What does it do btw?

Reply:
It returns the N most probable token along with the your generated token. (does not change results but simply returns more info)

Reply:
Ok thanks!